toolchain/base/value_ids.h

```diff
@@ -41,20 +50,147 @@ class Real : public Printable<Real> {
   bool is_decimal;
 };
 
-// Corresponds to an integer value represented by an APInt. This is used both
-// for integer literal tokens, which are unsigned and have an unspecified
-// bit-width, and integer values in SemIR, which have a signedness and bit-width
-// matching their type.
-struct IntId : public IdBase, public Printable<IntId> {
+// Corresponds to a canonicalized integer value. This is used both for integer
+// literal tokens, and integer values in SemIR. These always represent the
+// abstract mathematical value -- signed and regardless of the needed precision.
+//
+// Small values are internalized into the ID itself. Large values are
+// represented as an index into an array of `APInt`s with a canonicalized bit
+// width.
+class IntId : public Printable<IntId> {
+ public:
   using ValueType = llvm::APInt;
+
   static const IntId Invalid;
-  using IdBase::IdBase;
+
+  static auto MakeIndexOrInvalid(int index) -> IntId {
+    CARBON_DCHECK(index >= 0 && index <= InvalidIndex);
+    return IntId(ZeroIndexId - index);
```

Comment:
Is there validation that this doesn't produce incorrect values? Is it possible to have a unit test that tries making too many unique integers, to check for graceful failure?

Reply:
Sure, happy to look at testing the lexer limit in a follow-up.